docs/specifications/tests/Zone-TP/zone09.md

```diff
@@ -1,47 +1,276 @@
+Z09_NO_RESPONSE_MX_QUERY            | WARNING | ns_ip_list              | List of name servers that did not respond on MX query.
+Z09_NULL_MX_WITH_NON_ZERO_PREFERENCE| NOTICE  |                         | The child zone has a null MX with incorrect preference (must be 0).
+Z09_NULL_MX_WITH_OTHER_MX           | WARNING |                         | The child zone has a null MX but incorrectly mixed with other MX records.
+Z09_TLD_MAIL_DOMAIN_HARMFUL         | WARNING |                         | The child zone is a TLD and has MX, which is considered harmful.
+Z09_UNEXPECTED_RCODE_MX_RESPONSE    | WARNING | ns_ip_list, rcode       | The name servers listed returns unexpected RCODE value (listed) on the MX query.
+Z09_ROOT_MAIL_DOMAIN_MEANINGLESS    | NOTIFY  |                         | The root zone has an MX RRset, which is meaningless.
+
+The value in the Level column is the default severity level of the message. The
+severity level can be changed in the [Zonemaster-Engine profile]. Also see the
+[Severity Level Definitions] document.
+
+The argument names in the Arguments column lists the arguments used in the
+message. The argument names are defined in the [argument list].
+
+
+## Test procedure
+
+1.  Create an SOA query for the *Child Zone* without OPT record (no EDNS)
+    ("SOA Query").
+
+2.  Create an MX query for the *Child Zone* without OPT record (no EDNS)
+    ("MX Query").
+
+3.  Obtain the set of name server IP addresses using [Method4] and [Method5]
+    ("Name Server IP").
+
+4. Create the following empty sets
+
+    1.  Name server IP addresses ("No Response MX Query").
+    2.  Name server IP addresses and RCODE value ("Unexpected RCODE MX Query").
+    3.  Name server IP addresses ("Non-authoritative MX").
+    4.  Name server IP addresses ("No MX RRset").
+    5.  Name server IP addresses and MX RRset ("MX RRset").
 
-### Intercase dependencies
+5.  For each name server IP in *Name Server IP* do:
+
+    1. Send *SOA Query" over UDP to the name server.
+    2. Go to next name server IP if at least one of the following criteria is
+       met:
+       1. There is no DNS response.
+       2. The RCODE of the response is not "NoError" (IANA RCODE List).
+       3. The AA flag is not set in the response.
+       4. There is no SOA record with owner name matching the query.
+
+    2. Send *MX Query* over UDP to the name server and collect the
+       response, and:
+       1. If the response has the TC flag set, re-query over TCP and use that
+          response instead.
+       2. If there is no DNS response, then add the name server IP to the
+          *No Response MX Query* set.
+       3. Else, if the RCODE of response is not "NoError" ([IANA RCODE List]),
+          then add the name server IP and the RCODE to the
+          *Unexpected RCODE MX Query* set.
+       4. Else, if the AA flag is not set in the response, then add the name
+          server IP to the *Non-authoritative MX* set.
```

Comment:
This is maybe a silly question, how can we have the AA flag not set here but set in 1.3 (so that we can go on with the same name server)?

Reply:
It is unlikely. If we ignore it here, we could get strange results when it happens, but that is maybe OK. -- The zone could be served by a DNS "firewall" that responds on some queries but forwards others to the real name server for other queries. -- The zone could be served by two name servers behind a load balancer, and one of the name servers is misconfigured. We send two queries and they do not go to the same server.

Reply:
Fair enough. So this is just an extra check that might never catch anything. This is fine for me.